src/Storage/Redux/Slices/Discovery.ts

```diff
@@ -100,17 +162,70 @@ export const DiscoverySlice = createSlice({
             } else {
                 state.favorites.push(bookmark)
             }
+
+            // Add to favoriteRefs for all bookmark types (app-hub, vbd, custom)
+            if (!state.favoriteRefs) state.favoriteRefs = []
+            const order = state.favoriteRefs.length
+            state.favoriteRefs.push(createDAppReference(bookmark, order))
         },
         removeBookmark: (state, action: PayloadAction<{ href: string; isCustom?: boolean }>) => {
             const { isCustom, href } = action.payload
             if (isCustom) {
                 state.custom = state.custom.filter(dapp => !URIUtils.compareURLs(dapp.href, href))
-            } else {
-                state.favorites = state.favorites.filter(dapp => !URIUtils.compareURLs(dapp.href, href))
+                if (state.favoriteRefs) {
+                    state.favoriteRefs = state.favoriteRefs.filter(
+                        ref => ref.type !== "custom" || !URIUtils.compareURLs(ref.url, href),
+                    )
+                }
+                return
+            }
+
+            const favoriteMatch = state.favorites.find(dapp => URIUtils.compareURLs(dapp.href, href))
+            state.favorites = state.favorites.filter(dapp => !URIUtils.compareURLs(dapp.href, href))
+
+            if (state.favoriteRefs) {
+                const featuredMatch = state.featured.find(dapp => URIUtils.compareURLs(dapp.href, href))
+                const possibleAppHubId = featuredMatch?.id || DAppUtils.generateDAppId(href)
+                const possibleVbdId = favoriteMatch?.veBetterDaoId || featuredMatch?.veBetterDaoId
+
+                state.favoriteRefs = state.favoriteRefs.filter(ref => {
```

Comment:
I'd just check if it's the same ref type & the id matches, wdyt?

Reply:
Refactored.